core/src/main/scala/akka/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -96,61 +95,39 @@ import scala.util.control.NonFatal
   }
 
   private[KafkaConsumerActor] trait CommitRefreshing {
-    def add(offsets: Map[TopicPartition, OffsetAndMetadata]): Unit
-    def committed(offsets: java.util.Map[TopicPartition, OffsetAndMetadata]): Unit
-    def revoke(revokedTps: Set[TopicPartition]): Unit
-    def refreshOffsets: Map[TopicPartition, OffsetAndMetadata]
     def updateRefreshDeadlines(tps: Set[TopicPartition]): Unit
-    def assignedPositions(assignedTps: Set[TopicPartition], assignedOffsets: Map[TopicPartition, Long]): Unit
-    def assignedPositions(assignedTps: Set[TopicPartition],
-                          consumer: Consumer[_, _],
-                          positionTimeout: java.time.Duration): Unit
+    def refreshOffsets: Map[TopicPartition, OffsetAndMetadata]
   }
 
   private[KafkaConsumerActor] object CommitRefreshing {
-    def apply(commitRefreshInterval: Duration): CommitRefreshing =
+    def apply(commitRefreshInterval: Duration, progress: () => ConsumerProgressTracking): CommitRefreshing =
       commitRefreshInterval match {
-        case finite: FiniteDuration => new Impl(finite)
-        case _ => NoOp
+        case finite: FiniteDuration => new Impl(finite, progress())
+        case _ => new Noop()
       }
 
-    private object NoOp extends CommitRefreshing {
-      def add(offsets: Map[TopicPartition, OffsetAndMetadata]): Unit = {}
-      def committed(offsets: java.util.Map[TopicPartition, OffsetAndMetadata]): Unit = {}
-      def revoke(revokedTps: Set[TopicPartition]): Unit = {}
-      val refreshOffsets: Map[TopicPartition, OffsetAndMetadata] = Map.empty
-      def updateRefreshDeadlines(tps: Set[TopicPartition]): Unit = {}
-      def assignedPositions(assignedTps: Set[TopicPartition], assignedOffsets: Map[TopicPartition, Long]): Unit = {}
-      def assignedPositions(assignedTps: Set[TopicPartition],
-                            consumer: Consumer[_, _],
-                            positionTimeout: java.time.Duration): Unit = {}
+    private final class Noop() extends CommitRefreshing {
+      override def updateRefreshDeadlines(tps: Set[TopicPartition]): Unit = {}
+
+      override val refreshOffsets: Map[TopicPartition, OffsetAndMetadata] = Map()
     }
+    private final class Impl(commitRefreshInterval: FiniteDuration, progress: ConsumerProgressTracking)
+        extends CommitRefreshing
+        with ConsumerProgressTracking {
+      progress.addProgressTrackingCallback(this)
```

Comment:
This delegation gives me a hard time.

Reply:
My preference was to the looser coupling, rather than having some sort of wonky inheritance thing. Open to suggestions though!

Reply:
The constraints I came up with for the tracking were:
 * only a single object should track the requested/committed/assigned state for the consumer
 * other objects can query that state, but should not be making updates (outside of calls by the ConsumerActor)
 * commit refreshing needs to know about the changes in consumer state so it can update its internal state.

That's how i ended up with a parent tracker that calls listeners for calls to those update methods. Wanted to avoid another type in the hierarchy for the base (which listeners would implement) and then a separate one for the tracker that lets objects access the consumer progress state.

But yeah, its a bit weird that CommitRefreshing$Impl has committedOffsets/requestedOffsets. My expectation was that it is hidden away by the interface, so its not actually that big of a deal.

Reply:
As we implement two interfaces with similar things it becomes blurry which method is here for wich reason.

Maybe an anonymous class make that more clear?
```scala
      progress.addProgressTrackingCallback(new ConsumerProgressTracking {
        // required by the progress tracking trait, but not actually used or exposed as this type should only be
        //accessed by its CommitRefreshing interface
        override def committedOffsets: Map[TopicPartition, OffsetAndMetadata] = null
        override def requestedOffsets: Map[TopicPartition, OffsetAndMetadata] = null

        override def revoke(revokedTps: Set[TopicPartition]): Unit = {
          refreshDeadlines = refreshDeadlines -- revokedTps
        }

        override def assignedPositions(assignedTps: Set[TopicPartition],
                                       assignedOffsets: Map[TopicPartition, Long]): Unit = {
          // assigned the partitions, so update all the of deadlines
          refreshDeadlines = refreshDeadlines ++ assignedTps.map(_ -> commitRefreshInterval.fromNow)
        }
      })
```

Reply:
I'm having similar trouble understanding this type hierarchy.  I think it would make sense to have a separate type for the listener that is implemented outside this source file and seal the rest of traits.

Reply:
Which is sounds similar to what I have in the latest. But im the first to admit the weakness in my scala-fu.

Reply:
What I meant was in this type:

```
trait ConsumerProgressTrackingListener {
  def requested(offsets: Map[TopicPartition, OffsetAndMetadata]): Unit = {}
  def received[K, V](records: ConsumerRecords[K, V]): Unit = {}
  def committed(offsets: java.util.Map[TopicPartition, OffsetAndMetadata]): Unit = {}
  def revoke(revokedTps: Set[TopicPartition]): Unit = {}
  def assignedPositions(assignedTps: Set[TopicPartition], assignedOffsets: Map[TopicPartition, Long]): Unit = {}
}
```

We have default impls for the first 4 methods, but they're only ever implemented by `ConsumerProgressTracking`. Do they belong there instead?

Reply:
My thought was that if we are going to have a listener for progress, it should be able to listen to all the progress events, if it wants. Felt "lame" to only allow listening for a subset of events. That said, unless you need it YAGNI... i'll drop them. Easy enough to add later :)